`lib.fileset`: Representation for empty file sets without a base path

infinisil opened the pull request:
## Description of changes

Changes the internal fileset structure to also be able to represent empty file sets without any base path. This can then be used as the return value of `unions []`.

Once the `lib.fileset.intersect` combinator from the [full WIP combinators](https://github.com/NixOS/nixpkgs/pull/222981) is added, this will also be suitable as a return value for e.g. `intersect /foo /bar`.

Notably the new empty singleton value is not exposed in the interface, because I don't know of any use case for it. It could be added later if a use case emerges.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

## Things done
- [x] Added tests
- [x] Benchmarked (minor change, not significant)

infinisil committed:
lib.fileset: Internal representation v3, empty value without a base

`unions []` now works!

Notably the new empty value without a base is not exposed in the interface.
I don't know of any use case for it.

fricklerhandwerk commented on lib/fileset/default.nix:
```diff
@@ -170,7 +171,7 @@ If a directory does not recursively contain any file, it is omitted from the sto
             - If you want to import the file into the store _without_ a containing directory, use string interpolation or `builtins.path` instead of this function.
             - If you want to import the file into the store _with_ a containing directory, set `root` to the containing directory, such as ${toString (dirOf root)}, and set `fileset` to the file path.''
-    else if ! hasPrefix root fileset._internalBase then
+    else if ! fileset._internalIsEmptySingleton && ! hasPrefix root fileset._internalBase then
```
This will produce an awkward error message. Maybe check this separately?

infinisil replied:
I don't see how this would produce an awkward error message?

fricklerhandwerk commented on lib/fileset/default.nix:
```diff
@@ -156,7 +157,7 @@ If a directory does not recursively contain any file, it is omitted from the sto
     # Currently all Nix paths have the same filesystem root, but this could change in the future.
     # See also ../path/README.md
-    else if rootFilesystemRoot != filesetFilesystemRoot then
+    else if ! fileset._internalIsEmptySingleton && rootFilesystemRoot != filesetFilesystemRoot then
```
This will print `fileset: root "null"` in case of error, which won't be very helpful if you don't know there is such a thing as an empty file set without a base path.

infinisil replied:
Root is a path, it can never be null

fricklerhandwerk commented on lib/fileset/README.md:
```diff
@@ -41,9 +41,14 @@ An attribute set with these values:
+- `_internalVersion` (constant `3`, the current version):
   Version of the representation.
 
+- `_internalIsEmptySingleton` (bool):
+  Whether this file set is the singleton representing an empty file set.
```
```suggestion
- `_internalIsEmpty` (bool):
  Whether this file set is the empty file set.
```

Not sure if it's required to call it a singleton in the parameter. It's just *the* root-less empty set, regardless of how it's implemented.

infinisil replied:
I wanted to avoid calling it that because there's multiple ways to represent an empty fileset: `null` as the tree is another representation, `{ }` is another and there's many more if you have some files in the base path.

fricklerhandwerk replied:
What about this then?

```suggestion
- `_internalIsEmptyRootless` (bool):
  Whether this file set is the empty file set without a base path.
```

infinisil replied:
I like that, should be `_intetrnalIsEmptyBaseless` though. Will change

infinisil replied:
Although, it's not _just_ without a base path, it's also without a file set tree..

fricklerhandwerk replied:
Is it even possible to have a file set with a tree but without a base?

Also why isn't this property returned by a function? While it's all internal and users should not mess with it, the new field introduces another moving part and opportunity for error.

infinisil replied:
At least currently no base path also means no tree and the other way around, so yeah it might be okay to word it like that. In the end it doesn't matter much since indeed it is internal.

> Also why isn't this property returned by a function? While it's all internal and users should not mess with it, the new field introduces another moving part and opportunity for error.

It's not relevant for the user, it's an implementation detail. To the user it behaves just like they would expect. Exposing this as a function would expose implementation details.

fricklerhandwerk replied:
To be clear, that helper function would also be internal.

roberth replied:
Besides being returned for an empty union, isn't its unique property that its the identity of any union operation?
You may consider `_isUnionIdentity`.

fricklerhandwerk commented:
I'm not sure about the approach here. I understand that having a value to represent `unions []` would be nice to have, but the special case clutters the code quite a bit, moving it past the sweet spot of readability, and that particular edge case seems really not that important. Also I don't see why `intersect /foo /bar` can't return `{ base = /; tree` = null; }`. The overall change is therefore of questionable utility to me.

infinisil commented:
> Also I don't see why `intersect /foo /bar` can't return `{ base = /; tree` = null; }`.

The motivation comes from this example, which should work:

```nix
toSource {
  root = ./foo;
  fileset = intersect ./foo ./.;
}
```

If `intersect ./foo ./.` would use the common prefix (`./.`) as the base path, `toSource` would complain that the base path is outside of the root, when clearly that's not what the user would expect. So the base path of `intersect ./foo ./.` should instead be `./foo`.

Extending that, this should also succeed without an error:
```nix
toSource {
  root = ./foo;
  fileset = intersect ./foo ./bar;
}
```

though this would give an empty result, but that's as one would expect.

(this will definitely have to go into the design document)

fricklerhandwerk commented:
> So the base path of `intersect ./foo ./.` should instead be `./foo`.

Agreed that the combination of `intersect ./foo ./.` and `toSource { root = ./foo; }` as you describe it may be confusing, but not convinced about  `intersect` in isolation. Just from looking at it, it's not evident what the base path *should* be, and intuitively I'd say it's their common prefix. If you want it to be the longest path in the resulting set, it should be documented very visibly. Maybe even guard it by an invariant that the base path is never shortened in operations?

Only then it seems inevitable to produce a special kind of value for `intersect ./foo ./bar`.

@roberth: Fully agreed. A particularly involved implementation is still better with a well-articulated rationale.

infinisil commented:
> Just from looking at it, it's not evident what the base path _should_ be, and intuitively I'd say it's their common prefix. If you want it to be the longest path in the resulting set, it should be documented very visibly.

The base path is mostly an invisible implementation detail, users shouldn't have to know about it. The only interaction of it with the user is in this error message:

```
nix-repl> fileset.toSource { root = ./foo; fileset = ./.; }
error: lib.fileset.toSource: `fileset` could contain files in /home/user/src, which is not under the `root` (/home/user/src/foo). Potential solutions:
   - Set `root` to /home/user/src or any directory higher up. This changes the layout of the resulting store path.
   - Set `fileset` to a file set that cannot contain files outside the `root` (/home/user/src/foo). This could change the files included in the result.
```

This error should only be thrown when it's relevant, meaning when there could be some file outside the `root` in the `fileset`.

`intersect ./foo ./.` can never contain files outside `./foo`, so it should not thrown this error. Similarly, `intersect ./foo ./bar` can never contain any files at all, so it should not throw that error either.

infinisil commented:
> Maybe even guard it by an invariant that the base path is never shortened in operations?

Comparatively, `union` needs the common prefix path as the base path, it's a duality! `union ./foo ./bar` can contain files from either, so the base path should be `./.`.

infinisil commented:
I now updated this to use `_internalIsEmptyBaseless`, I liked this one the most. I'd rather not bikeshed over the name any longer, this is not exposed in the API so it can also be changed later if desired.

Furthermore, I added a section to the design document explaining why such a value is needed and the problem with alternatives. Valentin also confirmed in a meeting that he agrees with this being needed.

I think this is ready to be merged, and unless there are any complaints in the next couple days I'll merge this myself.

infinisil commented:
In PM's @fricklerhandwerk made the suggestion to call it `..withoutBase` or `..noBase` instead. That sounds good to me, so I now changed it to `_internalIsEmptyWithoutBase`.

roberth commented on lib/fileset/README.md:
```diff
@@ -41,9 +41,16 @@ An attribute set with these values:
   Tag to indicate this value is a file set.
 
-- `_internalVersion` (constant `2`, the current version):
+- `_internalVersion` (constant `3`, the current version):
```
Should this be in a file like `fileset/internal/README.md`?

infinisil replied:
I view `lib/fileset/README.md` as the internal contributor documentation, where I think this makes sense to have. It's not part of the public interface, but you need to know about it to contribute, which can mean to change the internals

roberth replied:
It should link to the user docs.

roberth reviewed (approved):
One off-topic suggestion. LGTM!
